cmd/promtool/main.go

```diff
@@ -303,9 +304,42 @@ func checkRules(filename string) (int, []error) {
 		numRules += len(rg.Rules)
 	}
 
+	dRules := checkDuplicates(rgs.Groups)
+	if len(dRules) != 0 {
+		fmt.Printf("%d duplicate rules(s) found.\n", len(dRules))
+		for _, n := range dRules {
+			fmt.Printf("Metric: %s\nLabel(s):\n", n.metric)
+			for i, l := range n.label {
+				fmt.Printf("\t%s: %s\n", i, l)
+			}
+		}
+		fmt.Println("Might cause inconsistency while recording expressions.")
+	}
+
 	return numRules, nil
 }
 
+type duplicateRules struct {
+	metric string
+	label  map[string]string
+}
+
+func checkDuplicates(r []rulefmt.RuleGroup) []duplicateRules {
+	var stack, stackDuplicates []duplicateRules
+
+	for _, rule := range r {
+		for _, props := range rule.Rules {
+			for _, inst := range stack {
+				if reflect.DeepEqual(inst, duplicateRules{metric: props.Record, label: props.Labels}) {
+					stackDuplicates = append(stackDuplicates, duplicateRules{metric: props.Record, label: props.Labels})
+				}
+			}
+			stack = append(stack, duplicateRules{metric: props.Record, label: props.Labels})
```

Comment:
Building this up is odd, I'd suggest using an index variable in the for.

Reply:
sure

Reply:
More exactly, get rid of the stack - you don't need it. Nested iteration over the rules will do it.

Reply:
@brian-brazil, stack array is used only to check whether the duplicate rule has been already scanned in the earlier iteration since we aim to print the duplicate rule only once in the warning. 

Please note that stack does not implement the stack concept. It is surely misleading and i will change in the following commit.

Suppose if I remove the `stack array` (which is used for maintaining a check) and do simple nesting, then we might print a duplicated rule twice in the output like in the case when the label value is repeated more than twice.

Please correct me if I am getting it wrong.

Reply:
@brian-brazil , done! pleae have a look.